### PR TITLE
Improve date parsing for our purposes to be less tolerant.

### PR DESCRIPTION
**Pull request by jonathanswenson**
Any timestamp that comes back from looker as a timestamp should be in iso8601 format. Therefore we should only parse timestamps that are in the valid format into datetimes.

Sawyer::Serializer automatically tries to convert any field that it believes could be a date into a date using Time.parse. Ruby's Time.parse is actually really lenient and will parse '30 days' into the 30th of the current month at midnight.
This becomes a problem when we parse certain attributes which automatically convert certain strings to dates when they shouldn't have been converted.

For example the json response from the server:

```
{
  "filters": {
    "users.created_date" : "30 days"
  },
  ...
}
```

would be parsed into a Sawyer::Resource with the following structure:

```
{
  "filters" => {
    "users.created_date" => 2016-08-30 00:00:00 -700
  },
  ...
}
```

Attempting to run the query that came back from the response (run_inline_query) would get very different data then running the query by id (run_query)

/cc @jbandhauer @dthorpe Thoughts?

### CODE REVIEW

**Comment by dthorpe** [comment]
Lock it down!  Approach sounds good.  Code looks good.

What's with the 2 travis runs? One pass, one fail.

**Comment by jbandhauer** [comment]
Someone originally set up travis here to do one run of the merged state and one of the branch as-it-is. 
The tests were failing because of the need for https://github.com/looker/looker-sdk-ruby/pull/47
I merged that change and restarted the tests here. Now the merge test works, but the original branch, of course, lacks that change. So, it still fails.

Also, I suggested a couple of changes to this PR to @jonathanswenson offline. So, expect another commit here. If he is so inclined, he could merge in master to the branch and both of these tests could be expected to succeed.

**Comment by dthorpe** [comment]
The larger issue here seems to be that we have filter expressions whose type is a superset of the type of the given field.  "30 days" is an expression that presumably evaluates to a date range, but users.created_date is just a date, so there is a type dissonance. That dissonance becomes an issue when something like Sawyer is expecting to validate or format the right-hand side value based on the observed type of the left-hand side.

Just making an observation.  No action item per se.

**Comment by jbandhauer** [comment]
Huh, we didn't merge this yet? Is there something wrong with it?

File: lib/looker-sdk/client.rb [hunk 1]
```diff
@@ -238,6 +238,32 @@ class Serializer < Sawyer::Serializer
       def encode(data)
         data.kind_of?(Faraday::UploadIO) ? data : super
       end
+
+      # slight modification to the base class' decode_has_value function to
```
Comment (jbandhauer) [inline lib/looker-sdk/client.rb]: `decode_has_value` => 'decode_hash_value'

File: lib/looker-sdk/client.rb [hunk 2]
```diff
@@ -238,6 +238,32 @@ class Serializer < Sawyer::Serializer
...
+      # less permissive when decoding time values.
+      #
+      # See https://github.com/looker/helltool/issues/22037 for more details
+      def decode_hash_value(key, value)
```
Comment (jbandhauer) [inline lib/looker-sdk/client.rb]: I'd like to see a github link to the exact method you are replacing here. That is very helpful as people upgrade stuff and decide if existing monkey patches are still appropriate or not.

File: lib/looker-sdk/client.rb [hunk 3]
```diff
@@ -238,6 +238,32 @@ class Serializer < Sawyer::Serializer
       def encode(data)
         data.kind_of?(Faraday::UploadIO) ? data : super
       end
+
+      # slight modification to the base class' decode_has_value function to
+      # less permissive when decoding time values.
+      #
+      # See https://github.com/looker/helltool/issues/22037 for more details
```
Comment (jbandhauer) [inline lib/looker-sdk/client.rb]: Link to private helltool issue is not appropriate here. Please remove it. Ideally, file a short issue in *this* public repo that has a succinct problem statement.

File: lib/looker-sdk/client.rb [hunk 4]
```diff
@@ -238,6 +238,32 @@ class Serializer < Sawyer::Serializer
...
+          end
+        elsif value.is_a?(Hash)
+          decode_hash(value)
+        elsif value.is_a?(Array)
+          value.map { |o| decode_hash_value(key, o) }
+        else
+          value
+        end
+      end
     end
```
Comment (jbandhauer) [inline lib/looker-sdk/client.rb]: Sad to not have some tests of this.

File: lib/looker-sdk/client.rb (hunk 2 above)
Comment (jbandhauer) [inline lib/looker-sdk/client.rb]: I see: https://github.com/lostisland/sawyer/blob/6eced4624217eb65d8c90be9d81b11c0bebd81e4/lib/sawyer/serializer.rb#L101

Given that your method is mostly like the original, I'd be inclined to change this to a snippet that used `alias` to remember the original, and then just did something more like...

```ruby
if time_field?(key, value) && value.is_a?(String)
  begin
    Time.iso8601(value)
  rescue ArgumentError
    value
  end
else
  original_decode_hash_value(key, value)
end
```

Comment (dthorpe) [inline lib/looker-sdk/client.rb]: Would `super` suffice here, instead of aliases?

**Comment by dthorpe** [comment]
Fixes #53

**Pull request merged**